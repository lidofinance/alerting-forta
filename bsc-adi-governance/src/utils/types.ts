import { FindingSeverity, FindingType, ethers } from 'forta-agent'
import { Log } from '@ethersproject/abstract-provider'

export type EventOfNotice = {
  address: string
  event: string
  alertId: string
  name: string
  description: CallableFunction
  severity: FindingSeverity
  type: FindingType
}

export type TransactionDto = {
  logs: Log[]
  to: string | null
  block: {
    timestamp: number
    number: number
  }
}

export type LogDescription = ethers.utils.LogDescription
