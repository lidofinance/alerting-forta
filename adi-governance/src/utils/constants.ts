// TODO: replace with mainnet address
export const CROSS_CHAIN_FORWARDER_ADDRESS = '0x9d8548963Fa0a9BE7C434cA482dd5b79E8062d3A'

export const ARAGON_AGENT_ADDRESS = '0x3e40d73eb977dc6a537af587d48316fee66e9c8c'
