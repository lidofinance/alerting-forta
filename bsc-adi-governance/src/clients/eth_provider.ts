import { TransactionResponse } from '@ethersproject/abstract-provider'
import { ethers } from 'forta-agent'
import * as E from 'fp-ts/Either'
import { retryAsync } from 'ts-retry'
import { NetworkError } from '../utils/finding.helpers'

const DELAY_IN_500MS = 500
const ATTEMPTS_5 = 5

export class ETHProvider {
  private readonly jsonRpcProvider: ethers.providers.JsonRpcProvider

  constructor(jsonRpcProvider: ethers.providers.JsonRpcProvider) {
    this.jsonRpcProvider = jsonRpcProvider
  }

  public async getStartedBlockForApp(argv: string[]): Promise<E.Either<Error, number>> {
    let latestBlockNumber: number = -1

    if (argv.includes('--block')) {
      latestBlockNumber = parseInt(argv[4])
    } else if (argv.includes('--range')) {
      latestBlockNumber = parseInt(argv[4].slice(0, argv[4].indexOf('.')))
    } else if (argv.includes('--tx')) {
      const txHash = argv[4]
      const tx = await this.getTransaction(txHash)
      if (E.isLeft(tx)) {
        return E.left(tx.left)
      }

      if (tx.right.blockNumber !== undefined) {
        latestBlockNumber = tx.right.blockNumber
      }
    }
    if (latestBlockNumber == -1) {
      try {
        latestBlockNumber = await this.jsonRpcProvider.getBlockNumber()
      } catch (e) {
        return E.left(new NetworkError(e, `Could not fetch latest block number`))
      }
    }

    return E.right(latestBlockNumber)
  }

  public async getTransaction(txHash: string): Promise<E.Either<Error, TransactionResponse>> {
    try {
      const out = await retryAsync<TransactionResponse>(
        async (): Promise<TransactionResponse> => {
          const tx = await this.jsonRpcProvider.getTransaction(txHash)

          if (!tx) {
            throw new Error(`Can't find transaction ${txHash}`)
          }

          if (tx.blockNumber === undefined) {
            throw new Error(`Transaction ${txHash} was not yet included into block`)
          }

          return tx
        },
        { delay: DELAY_IN_500MS, maxTry: ATTEMPTS_5 },
      )

      return E.right(out)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not call getTransaction`))
    }
  }
}
