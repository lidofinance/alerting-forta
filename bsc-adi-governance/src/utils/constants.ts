export const CROSS_CHAIN_EXECUTOR_ADDRESS = '0x8e5175d17f74d1d512de59b2f5d5a5d8177a123d'
export const CROSS_CHAIN_CONTROLLER_ADDRESS = '0x40c4464fca8cacd550c33b39d674fc257966022f'
