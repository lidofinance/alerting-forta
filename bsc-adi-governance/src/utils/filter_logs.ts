import { ethers } from 'forta-agent'
import { LogDescription, TransactionDto } from './types'

export function filterLogs(txEvent: TransactionDto, eventSignature: string, eventAddress: string): LogDescription[] {
  const iface = new ethers.utils.Interface([eventSignature])
  const filteredLogs: LogDescription[] = []

  for (const log of txEvent.logs) {
    if (log.address.toLowerCase() !== eventAddress.toLowerCase()) {
      continue
    }

    try {
      const logDescription = iface.parseLog(log)
      filteredLogs.push(logDescription)
    } catch (e) {
      // Only one from eventsOfNotice could be correct
      // Others - skipping
    }
  }

  return filteredLogs
}
