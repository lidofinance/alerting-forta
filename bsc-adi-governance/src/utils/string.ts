export function bscscanAddress(address: string): string {
  return `[${address}](https://bscscan.com/address/${address})`
}
