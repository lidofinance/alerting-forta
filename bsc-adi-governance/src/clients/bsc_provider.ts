import { TransactionResponse } from '@ethersproject/abstract-provider'
import { ethers } from 'forta-agent'
import { either as E } from 'fp-ts'
import { retryAsync } from 'ts-retry'
import { ICRossChainControllerClient } from '../services/cross-chain-controller/CrossChainController.srv'
import { NetworkError } from '../utils/errors'
import { BaseAdapter__factory, CrossChainController } from '../generated'
import { CROSS_CHAIN_CONTROLLER_ADDRESS } from '../utils/constants'

const DELAY_IN_500MS = 500
const ATTEMPTS_5 = 5
const MAINNET_CHAIN_ID = 1

export class BSCProvider implements ICRossChainControllerClient {
  private jsonRpcProvider: ethers.providers.JsonRpcProvider
  private crossChainControllerRunner: CrossChainController

  constructor(jsonRpcProvider: ethers.providers.JsonRpcProvider, crossChainControllerRunner: CrossChainController) {
    this.jsonRpcProvider = jsonRpcProvider
    this.crossChainControllerRunner = crossChainControllerRunner
  }

  public async getTransaction(txHash: string): Promise<E.Either<Error, TransactionResponse>> {
    try {
      const out = await retryAsync<TransactionResponse>(
        async (): Promise<TransactionResponse> => {
          const tx = await this.jsonRpcProvider.getTransaction(txHash)

          if (!tx) {
            throw new Error(`Can't find transaction ${txHash}`)
          }

          if (tx.blockNumber === undefined) {
            throw new Error(`Transaction ${txHash} was not yet included into block`)
          }

          return tx
        },
        { delay: DELAY_IN_500MS, maxTry: ATTEMPTS_5 },
      )

      return E.right(out)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not call getTransaction`))
    }
  }

  public async getBlockNumber(): Promise<E.Either<Error, number>> {
    try {
      const latestBlockNumber = await this.jsonRpcProvider.getBlockNumber()
      return E.right(latestBlockNumber)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not fetch latest block number`))
    }
  }

  public async getBridgeAdapters(): Promise<E.Either<Error, string[]>> {
    try {
      const bridgeAdapters = await this.crossChainControllerRunner.getReceiverBridgeAdaptersByChain(MAINNET_CHAIN_ID)
      return E.right(bridgeAdapters)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not get bridge adapters list from CrossChainController contract`))
    }
  }

  public async getBridgeAdaptersNamesMap(): Promise<E.Either<Error, Map<string, string>>> {
    try {
      const bridgeAdapters = await this.getBridgeAdapters()
      if (E.isLeft(bridgeAdapters)) {
        return E.left(bridgeAdapters.left)
      }
      const result = new Map<string, string>()

      for (const adapterAddress of bridgeAdapters.right) {
        const bridgeAdapterContract = BaseAdapter__factory.connect(adapterAddress, this.jsonRpcProvider)
        try {
          const adapterName = await bridgeAdapterContract.adapterName()
          result.set(adapterAddress, adapterName)
        } catch (error) {
          return E.left(new NetworkError(error, `Could not get adapter name from ${adapterAddress}`))
        }
      }

      return E.right(result)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not get bridge adapters list from CrossChainController contract`))
    }
  }

  public async getEnvelopeStateByIds(envelopeIds: string[]): Promise<E.Either<Error, Map<string, number>>> {
    try {
      const out = new Map<string, number>()
      await Promise.all(
        envelopeIds.map(async (envelopeId) => {
          const status = await this.crossChainControllerRunner['getEnvelopeState(bytes32)'](envelopeId)
          out.set(envelopeId, status)
        }),
      )
      return E.right(out)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not get bridge adapters list from CrossChainController contract`))
    }
  }

  public async getReceivedEnvelopeIds(fromBlock: number, toBlock: number): Promise<E.Either<Error, string[]>> {
    try {
      const results = await retryAsync(
        async () => {
          const keccakEvent = ethers.utils.keccak256(
            ethers.utils.toUtf8Bytes('TransactionReceived(bytes32,bytes32,uint256,(uint256,bytes),address,uint8)'),
          )
          const address = CROSS_CHAIN_CONTROLLER_ADDRESS
          return this.jsonRpcProvider.getLogs({
            fromBlock: `0x${fromBlock.toString(16)}`,
            toBlock: `0x${toBlock.toString(16)}`,
            address,
            topics: [keccakEvent, null, '0x0000000000000000000000000000000000000000000000000000000000000001', null],
          })
        },
        { delay: DELAY_IN_500MS, maxTry: ATTEMPTS_5 },
      )
      const iface = new ethers.utils.Interface([
        'event TransactionReceived(bytes32 transactionId, bytes32 indexed envelopeId, uint256 indexed originChainId, (uint256,bytes), address indexed bridgeAdapter, uint8 confirmations)',
      ])
      const out = results.map((result) => iface.parseLog(result))
      const res: string[] = out.map((out) => out.args.envelopeId)
      return E.right(res)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not call jsonRpcProvider.getStonksOrderParams`))
    }
  }
}
