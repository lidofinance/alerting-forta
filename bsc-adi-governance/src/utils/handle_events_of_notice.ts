import { Finding, FindingType } from 'forta-agent'
import { EventOfNotice, TransactionDto } from './types'
import { filterLogs } from './filter_logs'

export function handleEventsOfNotice(txEvent: TransactionDto, eventsOfNotice: EventOfNotice[]) {
  const findings: Finding[] = []

  for (const eventInfo of eventsOfNotice) {
    const filteredLogs = filterLogs(txEvent, eventInfo.event, eventInfo.address)
    for (const filteredLog of filteredLogs) {
      findings.push(
        Finding.fromObject({
          name: eventInfo.name,
          description: eventInfo.description(filteredLog.args),
          alertId: eventInfo.alertId,
          severity: eventInfo.severity,
          type: FindingType.Info,
          metadata: { args: String(filteredLog.args) },
        }),
      )
    }
  }

  return findings
}
