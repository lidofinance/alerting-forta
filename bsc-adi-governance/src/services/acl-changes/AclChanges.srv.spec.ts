import * as Winston from 'winston'
import { BlockEvent, Finding, getEthersProvider } from 'forta-agent'

import { AclChangesSrv } from './AclChanges.srv'
import { CrossChainController__factory } from '../../generated'
import { CROSS_CHAIN_CONTROLLER_ADDRESS } from '../../utils/constants'
import { BSCProvider } from '../../clients/bsc_provider'

const TEST_TIMEOUT = 120_000 // ms

describe('AclChanges srv functional tests', () => {
  const ethersProvider = getEthersProvider()

  const crossChainControllerRunner = CrossChainController__factory.connect(
    CROSS_CHAIN_CONTROLLER_ADDRESS,
    ethersProvider,
  )

  const bscClient = new BSCProvider(ethersProvider, crossChainControllerRunner)

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const service = new AclChangesSrv(logger, bscClient)

  let runBlock: (blockNumber: number, initBlock: number) => Promise<Finding[]>

  beforeAll(() => {
    jest.spyOn(console, 'log').mockImplementation(() => {})
    jest.spyOn(Date, 'now').mockImplementation(() => new Date('2024-01-01').getTime())

    runBlock = async (blockNumber, initBlock) => {
      if (initBlock > Number(blockNumber)) {
        throw new Error('Wrong initial block or argument order')
      }
      const BlockEventHack = { blockNumber } as BlockEvent
      return service.handleBlock(BlockEventHack)
    }
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  it(
    'should find owner not in whitelist',
    async () => {
      const findings = await runBlock(39776807, 39776806)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
})
