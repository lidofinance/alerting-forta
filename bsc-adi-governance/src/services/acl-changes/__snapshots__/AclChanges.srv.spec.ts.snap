// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AclChanges srv functional tests should find owner not in whitelist 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "BSC-ADI-VAULT-CONTRACT-OWNER",
    "description": "BSC ADI upgradeable Proxy Admin contract ([0x40c4464fca8cacd550c33b39d674fc257966022f](https://bscscan.com/address/0x40c4464fca8cacd550c33b39d674fc257966022f)) owner is set to contract address [0x250227c30e3b26525f292f63756e5eB3C425427E](https://bscscan.com/address/0x250227c30e3b26525f292f63756e5eB3C425427E)",
    "labels": [],
    "metadata": {
      "contract": "0x40c4464fca8cacd550c33b39d674fc257966022f",
      "name": "BSC ADI upgradeable Proxy Admin",
      "owner": "0x250227c30e3b26525f292f63756e5eB3C425427E",
    },
    "name": "ðŸš¨ BSC a.DI: Contract owner set to address not in whitelist",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 2,
    "uniqueKey": "",
  },
]
`;
